build-tools/automation/build.groovy

```diff
@@ -120,9 +122,17 @@ timestamps {
                 env.ghprbPullLongDescription = ''
 
                 if (hasPrLabel(env.GitRepo, env.ghprbPullId, 'full-mono-integration-build')) {
+                    hasPrLabelFullMonoIntegrationBuild = true
                     buildTarget = 'jenkins'
                 } else {
                     buildTarget = 'all'
+                    // Also compile host libs for windows so that a complete VSIX can be created
+                    if (isUnix()) {
+                        def uname = sh script: 'uname', returnStdout: true
+                        if (uname.startsWith("Darwin")) {
+                            EXTRA_MSBUILD_ARGS += " /p:AndroidSupportedHostJitAbis=Darwin:mxe-Win32:mxe-Win64"
```

Comment:
I'm still not sure I want to do this.

Which I suppose raises an interesting question: how many "different flavors" of packages do we want?  Right now (before this PR) I think we have 3-4 "flavors":

  * **Complete**: contains "everything" that isn't commercial; JITs for all supported ABIs, cross-compilers for all supported ABIs, etc.  This is what the [`xamarin-android`](https://jenkins.mono-project.com/view/Xamarin.Android/job/xamarin-android/) and [`xamarin-android-debug`](https://jenkins.mono-project.com/view/Xamarin.Android/job/xamarin-android-debug/) jobs produce.

  * **PR**: This is where things get weird.  Let's split this into three sub-sets:

      * **Bundled PR**: PR build which doesn't require a new mono, and thus can use a (presumably pre-existing) mono bundle.  Resulting packages will have JITs for all device ABIs and cross-compilers for all ABIs, for both macOS & Windows.

        Note that these will *not* (currently) contain Windows "host" `libmono-android.debug.dll` files, meaning Windows-side Xamarin.Android Designer integration tests cannot execute.

      * **New Mono PR**: PR build which requires a new mono bundle, and there is no pre-existing mono bundle.  Resulting packages will have JITs for a *subset* of device ABIs (see `$(AndroidSupportedTargetJitAbis)`; current `armeabi-v7a` and `x86`).  *No* cross-compilers will exist.  No LLVM will exist.  No Windows binaries will exist.

      * **`full-mono-integration-build` PR**: PRs which have the **full-mono-integration-build** label.  This *should* be identical to the **Complete** package.  (If it isn't, we should consider that a bug.)

In table form:

| Build Flavor | macOS libmono-android | macOS AOT & LLVM | Windows libmono-android | Windows AOT & LLVM |
| -------------|-----------------------|------------------|-------------------------|--------------------|
| Complete     |                   Yes |              Yes |                     Yes |                Yes |
| Bundled PR   |                   Yes |              Yes |                      No |                Yes |
| New Mono PR  |                   Yes |               No |                      No |                 No |

The unanswered question, though: *What do we want*?

*A* desire for PR builds is that they be Fast™¹, which has meant "don't build whatever isn't necessary to run the unit tests", in which case AOT-related tests are considered *optional*.

Thus the **New Mono PR** flavor, which builds the JIT (no AOT compilers, no LLVM), allowing us to execute Mono's BCL tests and `Mono.Android-Tests` without needing *yuuge* amounts of build time.

Unfortunately, into this existing world order we now have the Xamarin.Android Designer integration tests.  For macOS, this is Fine™; we always build a "host" mono & libmono-android, so the Designer tests should work on macOS.

Windows is another matter entirely; in order for the Designer to work, we need *both* a Windows `libmono` *and* a `libmono-android`.  Bundled PR builds would have the former but not the latter.

Hence this change, in which case macOS would *always* build Win32 and Win64 `libmono` and `libmono-android` files.

What we don't know: how impact will this have on PR times?  I doubt it'll be 0s; how much longer will build+test execution take with this change in place?

Is that time increase acceptable?

Is that time increase necessary?

If we *don't* do this `$(AndroidSupportedHostJitAbis)` change, it means we can *never* use PR build outputs to run Designer tests, *unless* the PR has the **full-mono-integration-build** label.  (Is *that* acceptable?)

We should still be able to run the Designer integration tests from *master* ("**Complete**") builds.

Short of build times showing that enabling WinAPI builds for all PRs results in a "negligible" or "acceptable" time increase, I'm wary of this particular change.

Meanwhile, if we keep the above table, this suggests another possible semantic change: should we "encode" the "build flavor" into the installer name?  If so, what encoding scheme should we use?

Spitballing, I think an "additive" encoding might be most useful, and thus e.g. `xamarin-android-oss-VERSION+AOT+designer` and `Xamarin.Android.Sdk-OSS-9.2.99.63_HEAD_4e1434c+AOT+designer.vsix`.

---

¹: ...but still useful² .
²: Which isn't defined.

Reply:
To raise another argument for these changes -- I believe that the `AndroidSupportedHostJitAbis` prop is also used to determine which `libzip.dll` files to build for Windows, which are required for these packages to be functional from a "system installation" standpoint.

Build time differences have been negligible in the results that I have seen locally, and also in a handful of pipeline PR builds while iterating on this. At best, our `build` step in the `xamarin-android-pr-pipeline` Jenkins job takes around 12 minutes and at worst it takes around 30. Here's a table of recent Jenkins PR build times: 

| Build Time | #2733 - Release | Others - Release | #2733 - Debug | Other - Debug |
| -------------|-----------------------|------------------|-------------------------|--------------------|
| Time 1     |    [build - (12 min in block)](https://jenkins.mono-project.com/view/Xamarin.Android/job/xamarin-android-pr-pipeline-release/100/flowGraphTable/)  |          [build - (10 min in block)](https://jenkins.mono-project.com/view/Xamarin.Android/job/xamarin-android-pr-pipeline-release/104/flowGraphTable/) |                   [build - (34 min in block)](https://jenkins.mono-project.com/view/Xamarin.Android/job/xamarin-android-pr-pipeline-debug/23/flowGraphTable/) |                [build - (28 min in block)](https://jenkins.mono-project.com/view/Xamarin.Android/job/xamarin-android-pr-pipeline-debug/19/flowGraphTable/)|
| Time 2  |       [build - (11 min in block)](https://jenkins.mono-project.com/view/Xamarin.Android/job/xamarin-android-pr-pipeline-release/103/flowGraphTable/) |              [build - (11 min in block)](https://jenkins.mono-project.com/view/Xamarin.Android/job/xamarin-android-pr-pipeline-release/108/flowGraphTable/) |                      |                |

Recent debug pipeline build results have been a bit harder to compare so far, and I am not sure if this is a result of differences in the pipeline as it is coming online, differences in hardware, or the fact that we're building more in debug by default? It seems that more recent [Debug pipeline PR builds](https://jenkins.mono-project.com/view/Xamarin.Android/job/xamarin-android-pr-pipeline-debug/) are again hovering around 10-12 minutes, so I am not entirely sure what caused the spike in the debug numbers above. In our "best case build timings" as shown above, the changes in this PR seem relatively negligible timing wise to me, especially compared to the time cost of our tests. 

I think further investment in breaking up the 2 hour test step so that various pieces can run on multiple machines simultaniously would offset any build time increases introduced by this PR. To that end, we will also be able to execute _more_ tests against PR build artifacts with these changes.